### DISCUSSION TITLE
Using usbipd and WSL2

### DISCUSSION THREAD

**Discussion by callimero** [SmingHub/Sming#2789] [General]
Hi,

I use usbipd to access my USB-Ports. (https://github.com/dorssel/usbipd-win)

So I have a /dev/ttyUSB0 in WSL2. 
My problem is that Sming still tries to use the powershell to upload and debug.

I can flash with no issues copying the flash command and use it when I omit the "powershell.exe"

I tried to find a configuration, but as I am a bit new here, I did not find a solution.

Ideas highly appreciated.
Carsten

**Comment by mikee47** [comment]
You can edit `Sming/Components/terminal/component.mk` and comment out the appropriate lines, or just change `ifdef WSL_ROOT` to `ifdef WSL_ROOTXXX` for example.

**Reply by callimero** [reply]
This will not affect all instances. 
I found undefine WSL_ROOT in build.mk is also ok and only is in one spot.

However, flashing is fine but Sming still tries to powershell.exe when opening a terminal after flashing. I am not sure why? Cant see why it is still defined in terminal/component.mk

Maybe it is also a good idea to support usbpid to make Sming for WSL2 even more Linux-like?

All the best,
Carsten

**Reply by mikee47** [reply]
The `TERMINAL` variable is cached, you need to run `make config-clean` to restore defaults as configured. If you can get usbpid working to your satisfaction maybe open a PR or issue with suggestions for improving support?

NB. I added all the WSL2 support shortly before abandoning Windows as a platform. Linux is just a better platform for this kind of development work!

**Reply by mikee47** [reply]
See https://sming.readthedocs.io/en/latest/getting-started/config.html#configuring-your-device regarding build variable caching.

**Reply by callimero** [reply]
WSL2 is really quite good! I am not quite through, with Sming (writing an article for the German Make Magazin) but this all works good, even the graphics simulator. And "code ." (VSCode) also as it looks. 

I will try further.
Carsten

**Reply by callimero** [reply]
> The `TERMINAL` variable is cached, you need to run `make config-clean` to restore defaults as configured. If you can get usbpid working to your satisfaction maybe open a PR or issue with suggestions for improving support?

That was it! A PR is a bit out of my comfort zone...  I guess an option WSL_USE_USBIPD or WSL_USE_POWERSHELL or similar would be enough. 

All the best,
Carsten

**Reply by mikee47** [reply]
@callimero I've raised PR #2790 which should provide a transparent fix for your situation. Could you give it a test to see if it works OK for you?

**Comment by slaff** [comment]
> but this all works good

@callimero not to mention that you can also use it together with Wokwi  https://sming.readthedocs.io/en/latest/experimental/wokwi.html ;)

**Reply by callimero** [reply]
We had already some wokwi articles in our magazin. Maybe a second article about Sming/Wokwi.
